### PR TITLE
Dynamic automatic answers system

### PR DESCRIPTION
**Pull request by Wonshtrum**
Make all automatic answers configurable.
The goal is to be able to provide an HTML-like file for each automatic answer generated by Sozu.
The format allows some templating to generate dynamic responses. The format is parsed from Http/1.1 to our Kawa intermediate format only once and at the start of a worker. Parsing resolves all templating variables and generates a trivial in-place replacement strategy. Filling a template should be as performant as possible.

### CODE REVIEW

File: lib/src/protocol/kawa_h1/answers.rs [hunk 1]
```diff
@@ -1,100 +1,609 @@
...
+            answer_502: Self::default_answer_502(),
+            answer_503: Self::default_answer_503(),
+            answer_504: Self::default_answer_504(),
+            answer_507: Self::default_answer_507(),
+        }
+    }
+}
+
+impl RawAnswers {
+    fn default_answer_301() -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (Keksoj) [reply lib/src/protocol/kawa_h1/answers.rs]: actually no

**Comment by Keksoj** [comment]
TODO:

- [ ] add status line to elements that can be passed to a Template
- [ ] list other elements that can be meaningful for each HTTP error kind

**Comment by Keksoj** [comment]
It would be meaningfull to add a `CLUSTER_ID` replaceable variable in the template, for errors 502, 503, 504 and 507.

**Comment by Keksoj** [comment]
providing cluster id as CLUSTER_ID where possible would be very nice too. So, to sum up:

- CLUSTER_ID

File: command/src/command.proto [hunk 2]
```diff
@@ -131,14 +129,13 @@ message HttpListenerConfig {
     required uint32 request_timeout = 10 [default = 10];
     // wether the listener is actively listening on its socket
     required bool active = 11 [default = false];
+    required CustomHttpAnswers http_answers = 12;
```
Comment (FlorentinDUBOIS) [inline command/src/command.proto:132]: Could we change the required keyword into optional one. The idea is to be able to deprecate and remove it in the future without a breaking change

Comment (Keksoj) [reply command/src/command.proto]: done!

File: command/src/command.proto [hunk 3]
```diff
@@ -164,6 +161,7 @@ message HttpsListenerConfig {
     // The tickets allow the client to resume a session. This protects the client
     // agains session tracking. Defaults to 4.
     required uint64 send_tls13_tickets = 20;
+    required CustomHttpAnswers http_answers = 21;
```
[the remark this answers is not in the document]
Comment (Keksoj) [reply command/src/command.proto]: done!

**Comment by Keksoj** [comment]
wait, I messed up something when rebasing main

File: lib/Cargo.toml [hunk 4]
```diff
@@ -34,7 +34,7 @@ hdrhistogram = "^7.5.4"
 hex = "^0.4.3"
 hpack = "^0.3.0"
 idna = "^0.5.0"
-kawa = { version = "^0.6.5", default-features = false }
+kawa = { git = "https://github.com/CleverCloud/kawa", default-features = false } # to replace with version 0.6.6 when available
```
Comment (FlorentinDUBOIS) [inline lib/Cargo.toml:37]: We need to publish the kawa v0.6.6 before going further :)

File: lib/src/protocol/kawa_h1/answers.rs [hunk 5]
```diff
@@ -1,100 +1,904 @@
...
-    pub PayloadTooLarge: Rc<Vec<u8>>,
-    /// 502
-    pub BadGateway: Rc<Vec<u8>>,
-    /// 503
-    pub ServiceUnavailable: Rc<Vec<u8>>,
-    /// 504
-    pub GatewayTimeout: Rc<Vec<u8>>,
-    /// 507
-    pub InsufficientStorage: Rc<Vec<u8>>,
+impl AsBuffer for SharedBuffer {
```
Comment (FlorentinDUBOIS) [inline lib/src/protocol/kawa_h1/answers.rs:16]: @Wonshtrum should we export a `AsReadOnlyBuffer` trait from kawa to remove the `panic` in implementation

File: lib/Cargo.toml (hunk 4 above)
Comment (Keksoj) [reply lib/Cargo.toml]: done!

**Review by FlorentinDUBOIS — APPROVED** [review]

**Pull request merged**